src/foam/core/ClassLoader.js

```diff
@@ -17,48 +17,92 @@
 
 foam.CLASS({
   package: 'foam.core',
+  name: 'ModelARequireExtension',
+  refines: 'foam.core.Model',
+
+  methods: [
+    function arequire() {
+      var p = Promise.resolve();
+      if (this.extends) {
+        p = p.then(this.__subContext__.arequire(this.extends));
+      }
+      for (var i = 0, a; a = this.axioms_[i]; i++) {
+        if (a.arequire) {
+          p = p.then(a.arequire());
+        }
+      };
+      return p;
+    },
+  ],
+});
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'RequiresARequireExtension',
+  refines: 'foam.core.Requires',
+
+  methods: [
+    function arequire() {
+      return this.__subContext__.arequire(this.path);
+    },
+  ],
+});
+
+foam.CLASS({
+  package: 'foam.core',
   name: 'ClassLoader',
 
   exports: [
     'arequire',
   ],
 
+  properties: [
+    {
+      name: 'pending',
+      class: 'Object',
+      factory: function() {
+        return {};
+      },
+    },
+  ],
+
   methods: [
     {
       name: 'arequire',
       class: 'foam.core.ContextMethod',
       code: function(X, modelId) {
-        // TODO: This method only adds requires. Make it add extends,
-        // properties, and anything else that might be needed.
+        if (this.__subContext__.isRegistered(modelId)) return Promise.resolve();
+        if (this.pending[modelId]) return this.pending[modelId];
```

Comment:
I think this works ok but maybe I didn't run the right test. I created a circular dependency in the models located in demos/node/js and ran both the node script and the browser at http://localhost:8000/?model=nodetooldemo.Test&classpath=/src/,/demos/node/js/ and it seemed to be fine.